sycl/test/devicelib/assert.cpp

```diff
@@ -1,6 +1,5 @@
 // REQUIRES: cpu,linux
-// RUN: %clangxx -fsycl -c %s -o %t.o
-// RUN: %clangxx -fsycl %t.o %sycl_libs_dir/libsycl-glibc.o -o %t.out
+// RUN: %clangxx -fsycl %s -lsycl -o %t.out
```

Comment:
> Link devicelib tests with libsycl instead of individual intermediate object files.

`%sycl_libs_dir/libsycl-glibc.o` is not an intermediate object file, see https://github.com/intel/llvm/blob/sycl/libdevice/cmake/modules/SYCLLibdevice.cmake#L33

These object files provide definitions for library functions used in device code. `libsycl.so` does not contain these definitions.

Reply:
Is that only for `libsycl-glibc.o` or also for the others ?

Reply:
That applies for all other object files used in sycl/test/devicelib directory:
- libsycl-msvc.o
- libsycl-glibc.o
- libsycl-complex.o
- libsycl-complex-fp64.o
- libsycl-cmath.o
- libsycl-cmath-fp64.o

Reply:
by the way, how do I build them ?
`make sycl-toolchain` does not do it, at least using my (possibly outdated) build recipe.

Reply:
@fwyzard python scripts should do the trick: https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#build-dpc-toolchain

Reply:
Or set LLVM_EXTERNAL_PROJECTS and LLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR: https://github.com/intel/llvm/blob/1d13f847309665c675b069a98f7524c291704f37/buildbot/configure.py#L72

Reply:
Is this how the upstream LLVM is being built ?
Otherwise it doesn't seem a particularly wise choice of an approach to me :-(

Anyway - since these "libdevice" object files are not part of `libsycl.so`, why do the test pass with the changes above ?

Reply:
> Is this how the upstream LLVM is being built ?
> Otherwise it doesn't seem a particularly wise choice of an approach to me :-(

Agree. I think these tests should be disabled if libdevice is not specified in LLVM_EXTERNAL_PROJECTS.
@vzakhari, FYI.

> Anyway - since these "libdevice" object files are not part of `libsycl.so`, why do the test pass with the changes above ?

They shouldn't. Make sure the test is not skipped by `REQUIRES: cpu,linux` (where `cpu` means OpenCL CPU Runtime).
Let's wait for the Buildbot sycl-ubu-x64-pr check to complete.

Reply:
Also, why do `make check-sycl` and `make check-sycl-cuda` try to run the `libdevice` tests even when it is not built ?

Reply:
I guess we'd better move `libdevice` tests into `libdevice` project directory.  Then, the `libdevice` tests that require `libsycl` may be enabled only if `libsycl` is actually being built.  There may still be pure `libdevice` tests that do not require `libsycl`, although, I do not think there are any right now.

@asavonic, @fwyzard, if it sounds reasonable to you, I can make this change.

Reply:
Either way should be fine: keeping the `libdevice` tests under sycl, and enabling them only if `libdevice` is built, or moving them to the `libdevice` directory, and enabling those that depend on SYCL only if it s being built.

Reply:
I still don't understand why I could get the tests to pass when `libdevice` was not built...

Reply:
> I guess we'd better move `libdevice` tests into `libdevice` project directory. Then, the `libdevice` tests that require `libsycl` may be enabled only if `libsycl` is actually being built. There may still be pure `libdevice` tests that do not require `libsycl`, although, I do not think there are any right now.
> 
> @asavonic, @fwyzard, if it sounds reasonable to you, I can make this change.

I think it is easier to propagate a flag from CMake to LIT and disable these tests if the feature is missing.

If we move them to the toplevel directory, we probably need to copy some bits from sycl/lit.cfg.py as well (all those CPU_RUN_PLACEHOLDER hacks). Not sure how to make them run for check-sycl as well.

Reply:
OK,  it is probably not worth it.